### DISCUSSION TITLE
Add depTypes for gradle

### DISCUSSION THREAD

**Discussion by cmplank** [renovatebot/renovate#12831] [Suggest an Idea]
For npm, I can setup an automerge rule using "matchDepTypes": "devDependencies". This is really handy for taking updates to libraries that I am confident won't impact the final artifact created by the build.

I would love to see the same type of feature for gradle. Gradle declares each dependency with `implementation`, `testImplementation`, or some others that are fairly rare in practice. The set of dependencies declared with `testImplementation` are similar in most ways to the `devDependencies` because they are only used in the test phase of the build and do not get packaged with the library.

Similarly, gradle has `plugin` dependencies. These are also generally build utilities where the versions wouldn't affect the final artifact that is created.

So I wanted to see what you all think or if you have better ways of managing automerge settings for your gradle projects.

**Comment by rarkins** [comment]
Are these standardised names or "free text"?

**Reply by cmplank** [reply]
I'm looking into it. They are called "dependency configurations" in Gradle. I had thought that they were standardized, but it looks like they come from plugins. Most notably, the java plugin includes several and these are what I'm familiar with (https://docs.gradle.org/current/userguide/java_plugin.html#sec:java_plugin_and_dependency_management).

Dependency Configurations are not free text either though. They are defined ahead of time in code (usually a plugin) so that each one actually behaves differently during the gradle build. The differences in behavior are around concepts like:

- whether needed for compilation or at runtime (or both)
- whether needed to compile tests
- whether needed on the classpath
- whether needed for annotation processing (a java thing)

I'm sure there are others and some must have nothing to do with java. I don't know anything about the Dependency Configurations created outside of the java plugin.

There is also the issue of inheritance since many of these DependencyConfigurations are extensions of others (for example `testImplementation` extends `implementation`). That's not something that Renovate would necessarily have to care about but it might create confusion for gradle users trying to use the matchDepTypes feature.

So it may not be a good candidate to add to renovate. Though Renovate could possibly set the depType for each dependency to the string value of the Dependency Configuration and leave it at that. That would be good enough for my needs, but I don't know if that might create more problems than it solves.

Gradle docs on dependency configurations: https://docs.gradle.org/current/userguide/declaring_dependencies.html

**Comment by trietsch** [comment]
An addition to this, would be to add the `plugin` keyword as well. This way, it would be possible to only match plugin dependencies for builds.

**Reply by simonbasle** [reply]
> An addition to this, would be to add the `plugin` keyword as well. This way, it would be possible to only match plugin dependencies for builds.

AFAIK this is already the case in `packageRules`:

```json
    {
      "matchDepTypes": [ "plugin" ],
      "matchUpdateTypes": [
        "minor",
        "patch"
      ],
      "groupName": "all non-major plugins",
      "groupSlug": "plugins-minor-patch"
    }
```

**Comment by scrocquesel** [comment]
Similar work was done for maven https://github.com/renovatebot/renovate/pull/13736, is there any similarity so it can converge eventually ?
